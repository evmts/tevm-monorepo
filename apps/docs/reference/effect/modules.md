[@evmts/effect](/reference/effect/README.md) / Exports

# @evmts/effect

## Table of contents

### Classes

- [CreateRequireError](/reference/effect/classes/CreateRequireError.md)
- [ParseJsonError](/reference/effect/classes/ParseJsonError.md)
- [RequireError](/reference/effect/classes/RequireError.md)

### Functions

- [createRequireEffect](/reference/effect/modules.md#createrequireeffect)
- [fileExists](/reference/effect/modules.md#fileexists)
- [logAllErrors](/reference/effect/modules.md#logallerrors)
- [parseJson](/reference/effect/modules.md#parsejson)

## Functions

### createRequireEffect

▸ **createRequireEffect**(`url`): `Effect`<`never`, [`CreateRequireError`](/reference/effect/classes/CreateRequireError.md), (`id`: `string`) => `Effect`<`never`, [`RequireError`](/reference/effect/classes/RequireError.md), `any`\>\>

An [Effect](https://www.effect.website/docs/introduction) wrapper around createRequire
createRequire is used to use the node.js `require` function in esm modules and cjs modules
in a way that is compatible with both. It also wraps them weith Effect for better error handling

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `url` | `string` | url to create require from |

#### Returns

`Effect`<`never`, [`CreateRequireError`](/reference/effect/classes/CreateRequireError.md), (`id`: `string`) => `Effect`<`never`, [`RequireError`](/reference/effect/classes/RequireError.md), `any`\>\>

require function

**`Example`**

```typescript
import { createRequireEffect } from '@eth-optimism/config'
const requireEffect = createRequireEffect(import.meta.url)
const solcEffect = requireEffect('solc')
```

**`See`**

https://nodejs.org/api/modules.html#modules_module_createrequire_filename

#### Defined in

[effect/src/createRequireEffect.js:50](https://github.com/evmts/evmts-monorepo/blob/main/effect/src/createRequireEffect.js#L50)

___

### fileExists

▸ **fileExists**(`path`): `Effect`<`never`, `never`, `boolean`\>

Checks if a file exists at the given path

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `path` | `string` | path to check |

#### Returns

`Effect`<`never`, `never`, `boolean`\>

true if the file exists, false otherwise

**`Example`**

```typescript
import { fileExists } from '@eth-optimism/config'
await fileExists('./someFile.txt')
```

#### Defined in

[effect/src/fileExists.js:16](https://github.com/evmts/evmts-monorepo/blob/main/effect/src/fileExists.js#L16)

___

### logAllErrors

▸ **logAllErrors**(`e`): `Effect`<`never`, `never`, `void`\>

Logs all errors and causes from effect

#### Parameters

| Name | Type |
| :------ | :------ |
| `e` | `unknown` |

#### Returns

`Effect`<`never`, `never`, `void`\>

**`Example`**

```typescript
import { logAllErrors } from '@eth-optimism/config'

someEffect.pipe(
  tapError(logAllErrors)
)

#### Defined in

[effect/src/logAllErrors.js:16](https://github.com/evmts/evmts-monorepo/blob/main/effect/src/logAllErrors.js#L16)

___

### parseJson

▸ **parseJson**(`jsonStr`): `Effect`<`never`, [`ParseJsonError`](/reference/effect/classes/ParseJsonError.md), { `compilerOptions?`: { `plugins`: { `name`: `string`  }[]  }  }\>

Parses a json string

#### Parameters

| Name | Type |
| :------ | :------ |
| `jsonStr` | `string` |

#### Returns

`Effect`<`never`, [`ParseJsonError`](/reference/effect/classes/ParseJsonError.md), { `compilerOptions?`: { `plugins`: { `name`: `string`  }[]  }  }\>

**`Throws`**

when the tsconfig.json file is not valid json

**`Example`**

```ts
const jsonEffect = parseJson('{ "compilerOptions": { "plugins": [{ "name": "@evmts/ts-plugin" }] } }')
````
@internal

#### Defined in

[effect/src/parseJson.js:33](https://github.com/evmts/evmts-monorepo/blob/main/effect/src/parseJson.js#L33)
