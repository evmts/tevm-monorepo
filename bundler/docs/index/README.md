**@tevm/bundler** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > index

# Module: index

## Index

### Type Aliases

- [AsyncBundlerResult](type-aliases/AsyncBundlerResult.md)
- [Bundler](type-aliases/Bundler.md)
- [BundlerResult](type-aliases/BundlerResult.md)
- [FileAccessObject](type-aliases/FileAccessObject.md)
- [Logger](type-aliases/Logger.md)
- [SolidityResolver](type-aliases/SolidityResolver.md)
- [SyncBundlerResult](type-aliases/SyncBundlerResult.md)

### Functions

- [bundler](functions/bundler.md)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
