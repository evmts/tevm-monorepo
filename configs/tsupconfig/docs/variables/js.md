**@tevm/tsupconfig** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > js

# Variable: js

> **`const`** **js**: `Options`

## Source

[js.js:3](https://github.com/evmts/tevm-monorepo/blob/main/configs/tsupconfig/src/js.js#L3)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
