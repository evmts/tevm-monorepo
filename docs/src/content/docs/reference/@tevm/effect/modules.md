---
editUrl: false
next: false
prev: false
title: "@tevm/effect"
---

## Modules

- [createRequireEffect](/reference/tevm/effect/createrequireeffect/readme/)
- [fileExists](/reference/tevm/effect/fileexists/readme/)
- [logAllErrors](/reference/tevm/effect/logallerrors/readme/)
- [parseJson](/reference/tevm/effect/parsejson/readme/)
- [resolve](/reference/tevm/effect/resolve/readme/)
