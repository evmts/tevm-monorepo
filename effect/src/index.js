export * from './createRequireEffect.js'
export * from './fileExists.js'
export * from './logAllErrors.js'
export * from './parseJson.js'
