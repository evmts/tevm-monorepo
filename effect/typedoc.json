{
	"$schema": "https://typedoc.org/schema.json",
	"out": "./docs",
	"entryPoints": ["./src/index.js"],
	"publicPath": "/reference/effect/",
	"plugin": ["typedoc-plugin-markdown"],
	"gitRevision": "main"
}
