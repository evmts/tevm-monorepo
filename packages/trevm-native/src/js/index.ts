import { createLogger } from '@tevm/logger';
import { InternalError } from '@tevm/errors';
import type { StateManager } from '@tevm/state';
import type { Common } from '@tevm/common';
import type { Chain } from '@tevm/blockchain';

// Import native module
import { init_trevm, TrevmWrapper } from '../../dist/index.node';

export interface TrevmEvmOptions {
  common: Common;
  stateManager: StateManager;
  blockchain: Chain;
  customPrecompiles?: any[];
  profiler?: boolean;
  allowUnlimitedContractSize?: boolean;
  loggingLevel?: string;
}

export interface TrevmRunCallOptions {
  caller?: string;
  to?: string;
  data?: string;
  value?: bigint;
  gasLimit?: bigint;
  skipBalance?: boolean;
}

export interface TrevmRunCallResult {
  returnValue: string;
  executionGasUsed: string;
  error?: string;
}

export interface TrevmEvm {
  runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult>;
}

/**
 * Creates a trevm-based EVM instance
 * @param options EVM options
 * @returns A trevm EVM instance
 */
export const createTrevm = async ({
  common,
  stateManager,
  blockchain,
  customPrecompiles,
  profiler,
  allowUnlimitedContractSize,
  loggingLevel,
}: TrevmEvmOptions): Promise<TrevmEvm> => {
  const logger = createLogger({
    name: '@tevm/trevm',
    level: loggingLevel ?? 'warn',
  });
  
  logger.debug({
    allowUnlimitedContractSize,
    profiler,
    customPrecompiles: customPrecompiles?.map(c => c.address.toString()),
  });

  try {
    // Initialize the native trevm module with callbacks to the StateManager
    const nativeTrevm = await init_trevm({
      chainId: common.chainId(),
      allowUnlimitedContractSize: allowUnlimitedContractSize ?? false,
      stateCallbacks: {
        getAccount: async (address: string) => {
          try {
            const account = await stateManager.getAccount(address);
            if (!account) return null;
            
            return {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? `0x${Buffer.from(account.codeHash).toString('hex')}` : undefined,
              code: account.code ? `0x${Buffer.from(account.code).toString('hex')}` : undefined,
            };
          } catch (error) {
            logger.error(`Error in getAccount for ${address}:`, error);
            return null;
          }
        },
        putAccount: async (address: string, account: any) => {
          try {
            await stateManager.putAccount(address, {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? Buffer.from(account.codeHash.slice(2), 'hex') : undefined,
              code: account.code ? Buffer.from(account.code.slice(2), 'hex') : undefined,
            });
            return true;
          } catch (error) {
            logger.error(`Error in putAccount for ${address}:`, error);
            return false;
          }
        },
        getStorage: async (address: string, key: string) => {
          try {
            const value = await stateManager.getContractStorage(address, key);
            return value ? `0x${Buffer.from(value).toString('hex')}` : '0x0';
          } catch (error) {
            logger.error(`Error in getStorage for ${address} key ${key}:`, error);
            return '0x0';
          }
        },
        putStorage: async (address: string, key: string, value: string) => {
          try {
            await stateManager.putContractStorage(
              address,
              key,
              Buffer.from(value.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putStorage for ${address} key ${key}:`, error);
            return false;
          }
        },
        getCode: async (address: string) => {
          try {
            const code = await stateManager.getContractCode(address);
            return code ? `0x${Buffer.from(code).toString('hex')}` : '0x';
          } catch (error) {
            logger.error(`Error in getCode for ${address}:`, error);
            return '0x';
          }
        },
        putCode: async (address: string, code: string) => {
          try {
            await stateManager.putContractCode(
              address,
              Buffer.from(code.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putCode for ${address}:`, error);
            return false;
          }
        },
        getBlockHash: async (blockNumber: number) => {
          try {
            const block = await blockchain.getBlock(BigInt(blockNumber));
            return block ? `0x${Buffer.from(block.header.hash()).toString('hex')}` : null;
          } catch (error) {
            logger.error(`Error in getBlockHash for block ${blockNumber}:`, error);
            return null;
          }
        }
      }
    });

    // Return a JavaScript wrapper around the native trevm that matches the Evm interface
    return {
      async runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult> {
        try {
          return await nativeTrevm.run_call({
            caller: opts.caller,
            to: opts.to,
            data: opts.data,
            value: opts.value,
            gas_limit: opts.gasLimit,
            skip_balance: opts.skipBalance,
          });
        } catch (error) {
          logger.error('Error in runCall:', error);
          throw new InternalError(`Trevm execution error: ${error}`);
        }
      }
    };
  } catch (error) {
    logger.error('Error initializing trevm:', error);
    throw new InternalError(`Failed to initialize trevm: ${error}`);
  }
};