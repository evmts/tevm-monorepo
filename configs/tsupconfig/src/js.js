import { createTsUpOptions } from './createTsupOptions.js'

export const js = createTsUpOptions({
	target: 'js',
})
