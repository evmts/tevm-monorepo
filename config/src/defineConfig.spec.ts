import { type CompilerConfig, defaultConfig, defineConfig } from './index.js'
import { execSync } from 'child_process'
import { runSync } from 'effect/Effect'
import {
	type MockedFunction,
	beforeEach,
	describe,
	expect,
	it,
	vi,
} from 'vitest'

vi.mock('child_process', () => ({
	execSync: vi.fn(),
}))

const mockExecSync = execSync as MockedFunction<typeof execSync>

describe(defineConfig.name, () => {
	beforeEach(() => {
		mockExecSync.mockReset()
	})

	it('should return a valid config when no config is provided', () => {
		const configFactory = () => ({})
		const config = defineConfig(configFactory as any)
		const resolvedConfig = config.configFn('./')
		expect(runSync(resolvedConfig)).toEqual(defaultConfig)
	})

	it('should run forge config --json command when foundryProject is provided', () => {
		const forgeCommandOutput = JSON.stringify({
			solc_version: '0.8.4',
			remappings: [],
			libs: ['lib1', 'lib2'],
		})
		mockExecSync.mockReturnValueOnce(Buffer.from(forgeCommandOutput))

		const configFactory = () =>
			({
				foundryProject: 'forge',
			}) as CompilerConfig
		const config = defineConfig(configFactory)
		const resolvedConfig = config.configFn('./')

		expect(runSync(resolvedConfig)).toEqual({
			remappings: defaultConfig.remappings,
			foundryProject: 'forge',
			libs: ['lib1', 'lib2'],
		})
		expect(mockExecSync).toHaveBeenCalledWith('forge config --json', {
			cwd: './',
		})
	})

	it('should default forge command to forge', () => {
		const forgeCommandOutput = JSON.stringify({
			solc_version: '0.8.4',
			remappings: [],
			libs: ['lib1', 'lib2'],
		})
		mockExecSync.mockReturnValueOnce(Buffer.from(forgeCommandOutput))

		const configFactory = () => ({
			foundryProject: true,
		})
		const config = defineConfig(configFactory)
		const resolvedConfig = config.configFn('./')

		expect(runSync(resolvedConfig)).toEqual({
			remappings: defaultConfig.remappings,
			foundryProject: true,
			libs: ['lib1', 'lib2'],
		})
		expect(mockExecSync).toHaveBeenCalledWith('forge config --json', {
			cwd: './',
		})
	})

	it('should throw if invalid config', () => {
		const configFactory = () =>
			({
				notACorrectKey: true,
			}) as CompilerConfig
		const config = defineConfig(configFactory)
		expect(() => runSync(config.configFn('./'))).toThrowErrorMatchingSnapshot()
	})

	it('should throw error when forge command fails', () => {
		mockExecSync.mockImplementationOnce(() => {
			throw new Error()
		})

		const configFactory = () => ({
			foundryProject: 'forge',
		})
		const config = defineConfig(configFactory)

		expect(() => runSync(config.configFn('./'))).toThrowErrorMatchingSnapshot()
	})

	it('should throw error when forge command output is not valid JSON', () => {
		mockExecSync.mockReturnValue('{"invalid JSON`{')

		const configFactory = () => ({
			foundryProject: 'forge',
		})
		const config = defineConfig(configFactory)

		expect(() => runSync(config.configFn('./'))).toThrowErrorMatchingSnapshot()
	})

	it('should throw error when forge remappings format is incorrect', () => {
		const forgeCommandOutput = JSON.stringify({
			remappings: ['invalid=remapping=format'],
		})
		mockExecSync.mockReturnValueOnce(Buffer.from(forgeCommandOutput))

		const configFactory = () => ({
			foundryProject: 'forge',
		})
		const config = defineConfig(configFactory)

		expect(() => runSync(config.configFn('./'))).toThrowErrorMatchingSnapshot()
	})

	it('should resolve remappings to absolute paths', () => {
		const forgeCommandOutput = JSON.stringify({
			remappings: ['key=value'],
		})
		mockExecSync.mockReturnValueOnce(Buffer.from(forgeCommandOutput))

		const configFactory = () => ({
			foundryProject: 'forge',
		})
		const config = defineConfig(configFactory)
		const resolvedConfig = config.configFn('/config')

		const res = runSync(resolvedConfig).remappings

		expect(mockExecSync).toBeCalledTimes(1)

		expect(res).toEqual({
			key: '/config/value',
		})
	})
})
