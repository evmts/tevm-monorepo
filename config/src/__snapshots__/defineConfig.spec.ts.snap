// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`defineConfig > should throw error when forge command fails 1`] = `
"FoundryNotFoundError: Unable to resolve EVMts CompilerConfig at ./
Failed to resolve forge config using \\"forge config --json\\" command. Make sure forge is installed and accessible and forge config --json works.
note: forge is used to fetch remappings only if forgeConfig is set. If you would prefer to not use forge you can set remappings
or lib directly in your EVMts compiler config and then EVMts will run without forge"
`;

exports[`defineConfig > should throw error when forge command output is not valid JSON 1`] = `
"FoundryConfigError: Unable to resolve EVMts CompilerConfig at ./
Unable to resolve foundry config using forge config --json"
`;

exports[`defineConfig > should throw error when forge remappings format is incorrect 1`] = `
"InvalidRemappingsError: Unable to resolve EVMts CompilerConfig at ./
Invalid remappings: invalid=remapping=format"
`;

exports[`defineConfig > should throw if invalid config 1`] = `
"InvalidConfigError: Unable to resolve EVMts CompilerConfig at ./
Invalid EVMts CompilerConfig detected"
`;
